### PR TITLE
Slice and dice inverse

### LINKED ISSUE #115: [OBJECTIVE]: Implement the inverse projection (planar → geodetic)
## Description
The current VGC projection is forward-only: it maps geodetic coordinates to a point on the icosahedron face plane. The inverse  is not yet implemented. This is required for any operation that starts from a ZONE index and needs to recover real-world coordinates: cell centroid, boundary vertices, zone-from-point, and the JS bindings' geometry output.
The inverse pipeline is the reverse of the forward one:
Planar → Barycentric → Cartesian → Authalic → Geodetic
Each step is a conversion between adjacent coordinate systems, so this issue depends directly on the CoordinateSystem trait (#103) having all conversion pairs implemented in the inverse direction.

## Acceptance Criteria
- [ ] cartesian_to_geo is implemented
- [ ] The inverse uses the CoordinateSystem conversion chain from #103 — no ad-hoc math duplicated in vgc.rs
- [ ] Roundtrip test: forward → inverse recovers the input geodetic coordinates within 1e-9° tolerance across all 20 icosahedron faces
- [ ] Edge cases tested: points near face edges, near face centroids, and near the icosahedron vertices
- [ ] The implementation handles the reoriented icosahedron from #102 (correct face assignment for the optimal DGGS orientation)

## Depends On
#102 #103 #104

### PR DESCRIPTION
**Pull request by joaocsmanuel**
Steps:
- [x] Link PR to the issue or kanban board item.
- [x] Write a list of what was done, preferably by bullet points.
- [ ] Add README documentation of what's done in the PR, if needed.
- [x] Request review

Core feature: implemented the inverse VGC projection (cartesian_to_geo) - previously just a stub on master. This was the main point of the branch, and involved fixing several bugs found along the way:

1. Face index off-by-one: ForwardCartesian.face is 1-indexed (matches gp-dggrs convention), but the inverse was using it directly as a 0-indexed polyhedron face lookup, corrupting the face template and 3D triangle used for every point. This will properly be fixed in another branch, for now its a temporary fix.
3. Wrong xy/uv recovery: the affine-inverse math recovering the sub-triangle interpolation parameters had a missing reciprocal (xy = t instead of xy = 1/t).
4. Wrong 3D point reconstruction: replaced a broken dot-product-based shortcut with exact spherical trilateration (solving for the point given true angular distances from two known triangle vertices, disambiguating the two mirror solutions using the triangle's center).
5. Broken Newton solver (inverse_slice_and_dice): its naive linear initial guess could land outside the valid parameter region and silently fail on the first iteration; fixed by bootstrapping from an approximate reconstruction instead.
6. Removed leftover debug println!s from both the forward and inverse paths.

Result: round-trip accuracy of ~1e-8° (sub-millimeter) verified across 9 globally-distributed test points (poles, antimeridian, multiple continents), added as a permanent regression test (test_roundtrip_many_points).

Supporting API changes:
- AuthalicSphere (in src/ellipsoid/authalic.rs) gained the reverse conversion: renamed convert → to_authalic and added to_geodetic (authalic → geodetic), so cartesian_to_geo can convert its reconstructed authalic coordinates back to geodetic degrees.
- Updated all call sites (vgc.rs, basic.rs, polygon.rs) and test helpers to match.

Examples:
- basic.rs: fixed to use the renamed AuthalicSphere API; fixed its round-trip demo, which was passing a raw Point instead of converting through AuthalicSphere first.
- inverse.rs: new example dedicated to demonstrating the forward→inverse round-trip; registered in Cargo.toml (this crate declares examples explicitly rather than relying on auto-discovery, since the directory is named example/ not the Cargo-conventional examples/).
- polygon.rs: minor call-site fix for the same rename.

Not part of this branch's work, pre-existing on master: test_pole_behavior and test_project_forward fail regardless of these changes (confirmed via git stash/baseline comparison at the start of this work) — unrelated forward-projection issues, out of scope here.

NOTE: Some files from viewer_3d come attached. I must have messed the rebase somewhow.

### CODE REVIEW

File: gp-proj/src/projections/projections/traits.rs [hunk 1]
```diff
@@ -22,6 +24,7 @@ pub struct ForwardBary {
 pub struct ForwardCartesian {
     pub coords: Coord,
     pub face: usize,
+    pub sub_triangle_id: u8,
```
Comment (sunayana) [inline gp-proj/src/projections/projections/traits.rs:25]: This makes sense to me. It is the right design choice here, the inverse needs to know which of the 6 sub-triangles a point landed in to invert!

File: gp-proj/src/projections/projections/traits.rs [hunk 2]
```diff
@@ -39,14 +42,13 @@ pub trait Projection {
         polyhedron: Option<&Polyhedron>,
         layout: Option<&dyn Layout>,
     ) -> Vec<ForwardCartesian>;
-    fn cartesian_to_geo(&self, coords: Vec<Coord>) -> Point;
-
-    fn compute_distortion(
+    fn cartesian_to_geo(
```
Comment (sunayana) [inline gp-proj/src/projections/projections/traits.rs:43]: This also seems correct and consistent with the forward direction's shape. This is a breaking API change though, as far as I see there is only one implementor right now!

File: gp-proj/src/projections/projections/vgc.rs [hunk 3]
```diff
@@ -154,6 +156,7 @@ impl Projection for Vgc {
                             y: p_y_face * r,
                         },
                         face: index + 1,
+                        sub_triangle_id,
```
Comment (sunayana) [inline gp-proj/src/projections/projections/vgc.rs:156]: this line 158 would now need to change to `face: index` right ? Since we work with 0-index now?

Comment (sunayana) [reply gp-proj/src/projections/projections/vgc.rs:156]: I think the tests `test_project_forward/test_pole_behavior` will pass after that fix.

File: gp-proj/src/projections/projections/vgc.rs [hunk 4]
```diff
@@ -164,23 +167,108 @@ impl Projection for Vgc {
...
+        polyhedron: Option<&Polyhedron>,
+    ) -> Vec<Point> {
+        let mut out: Vec<Point> = vec![];
+        let polyhedron = polyhedron.unwrap();
+        let sphere = AuthalicSphere::from_ellipsoid(&WGS84);
+        let r = self.radius;
+
+        for position in positions {
+            // ForwardCartesian.face is 1-indexed (see geo_to_cartesian), polyhedron APIs are 0-indexed
+            let face = usize::from(position.face) - 1;
```
Comment (sunayana) [inline gp-proj/src/projections/projections/vgc.rs]: I think line 181 can be dropped and the line 182 `-1` can also be dropped.

File: gp-proj/src/projections/projections/vgc.rs [hunk 5]
```diff
@@ -164,23 +167,108 @@ impl Projection for Vgc {
...
+            let ac_y = c.1 - a.1;
+
+            // B + t*(P-B) = A + s*(C-A)
+            // t*bp_x - s*ac_x = a.0 - b.0
+            // t*bp_y - s*ac_y = a.1 - b.1
+            let rhs_x = a.0 - b.0;
+            let rhs_y = a.1 - b.1;
+            let det = bp_x * (-ac_y) - bp_y * (-ac_x);
+            // t is the parameter such that D = B + t*(P-B); since P = B + xy*(D-B), t = 1/xy
+            let t = (rhs_x * (-ac_y) - rhs_y * (-ac_x)) / det;
```
Comment (sunayana) [inline gp-proj/src/projections/projections/vgc.rs]: Here `det ==0.0` when `P` lands on sub-triangle corner `B` and `t == 0.0` makes `1.0/t` a second `NaN` source. Needs an explicit ` if det.abs() < EPS { /* P == B, xy = 0 exactly */ }` short-circuit before the divide and similarly a guard `t == 0.0`.

File: gp-proj/src/projections/projections/vgc.rs [hunk 6]
```diff
@@ -334,6 +428,142 @@ fn affine_transform_triangle(
...
+        let d_uv_ap = (uv_ap - uv) / h;
+        let d_xy_bp = (xy_bp - xy) / h;
+        let d_uv_bp = (uv_bp - uv) / h;
+
+        // 2x2 Jacobian:
+        // | d_xy_ap  d_xy_bp | | delta_ap |   | r_xy |
+        // | d_uv_ap  d_uv_bp | | delta_bp | = | r_uv |
+
+        let det = d_xy_ap * d_uv_bp - d_xy_bp * d_uv_ap;
+        if det.abs() < 1e-14 {
```
Comment (sunayana) [inline gp-proj/src/projections/projections/vgc.rs:623]: Function returns `[f64; 2]` unconditionally; a caller has no way to know whether it broke due to convergence, singular Jacobian, or exhausting 50 iterations. Suggest returning `Result<[f64; 2], InverseError>` or at minimum a `converged: bool` alongside the value, since `cartesian_to_geo` currently trusts this blindly.

File: gp-proj/src/projections/projections/vgc.rs [hunk 7]
```diff
@@ -334,6 +428,142 @@ fn affine_transform_triangle(
...
+            break; // singular, can't continue
+        }
+
+        let delta_ap = (r_xy * d_uv_bp - r_uv * d_xy_bp) / det;
+        let delta_bp = (d_xy_ap * r_uv - d_uv_ap * r_xy) / det;
+
+        ap = (ap - delta_ap).clamp(0.0, ab);
```
Comment (sunayana) [inline gp-proj/src/projections/projections/vgc.rs]: Are you sure that the `.clamp` does not silently absorb a divergent step rather than signalling it?

**Review by sunayana — COMMENTED** [review]
@joaocsmanuel Nice work getting this far!
I went through this in detail  (diff, math, and a couple of empirical round-trip checks). 
The core approach feels sound to me, `reconstruct_point`'s spherical trilateration is an exact closed-form inverse of the geometry, `authalic.rs`'s new `to_geodetic` uses the correct inverse coefficient set, and `test_roundtrip_many_points` confirms round-tripping works to `~1e-8` rad across globally distributed points including near the poles.

Overall there are two blocking issues:
- face indexing and
- NaN at sub-triangle corners.

File: gp-proj/src/projections/projections/vgc.rs (hunk 3 above)
Comment (joaocsmanuel) [reply gp-proj/src/projections/projections/vgc.rs:156]: Yes, I will change this.

File: gp-proj/src/projections/projections/vgc.rs (hunk 5 above)
Comment (joaocsmanuel) [reply gp-proj/src/projections/projections/vgc.rs]: Ops, what a mistake. Nice catch. Fixed.

File: gp-proj/src/projections/projections/vgc.rs (hunk 6 above)
Comment (joaocsmanuel) [reply gp-proj/src/projections/projections/vgc.rs:623]: Good point. I will fix.

File: gp-proj/src/projections/projections/vgc.rs (hunk 7 above)
Comment (joaocsmanuel) [reply gp-proj/src/projections/projections/vgc.rs]: It wsa silently absorbing. Fixed.

**Comment by joaocsmanuel** [comment]
Check my latest commit @sunayana : https://github.com/GeoPlegma/GeoPlegma/pull/150/changes/864eff88f38007aa7cc2efc220041df3bbedb518 with the latest changes.

**Comment by joaocsmanuel** [comment]
I really want to merge this, I dont want clutter with more conflicts. Let me know @sunayana please

**Comment by sunayana** [comment]
Hi João,
I hope you received my DM. I have been sick with a heavy viral fever for
the past couple of weeks and still recovering. I don't mean to delay your
PR. So I'll go ahead and approve it right away if that works for you, so
that you are able to merge and I can revisit the changes once I am feeling
better.

**Review by sunayana — APPROVED** [review]
